### ISSUE TITLE
`~/.gitconfig` overrides `.gitpublish`

### ISSUE THREAD

**Issue by ehabkost** [stefanha/git-publish#52]
Currently the settings on repository-provided `.gitpublish` file can't override settings from `~/.gitconfig` because `~/.gitconfig` has a higher priority.

I'm not sure yet if this should be considered a bug. The current behavior may be expected by users that want to override `.gitpublish` options in their `~/.gitconfig`. Other users (myself included) may want their `~/.gitconfig` files to hold only defaults for repositories that don't have a `.gitpublish` file.

I'm opening this issue so we discuss what to do. I don't mind keeping the existing behavior by now, but I think we can support both use cases with the right combination of configuration options. I will try to come up with a proposal later.

**Comment by stefanha** [comment]
I think ~/.gitpublish should be lower priority than .gitpublish for two reasons:
1. To override user defaults on a per-project basis.
2. Consistency with git-config(1).

I'm thinking about whether we can somehow combine .git/config into this.  That way per-project settings that should be shared with other developers can be committed via .gitpublish but they can be overridden with .git/config (it's not committed to the repo).

**Comment by jnsnow** [comment]
For many moons, I forgot exactly why I raised this issue. Today, I hit it again and remembered. When a repo provides a .gitpublish with a "default" profile, there's no way to override these, because your own default profile in both .git/config and ~/.gitconfig are bypassed entirely.

I believe that `git_config_with_profile` ought to look in this order instead:
1. `.git/config`
2. `.gitpublish`
3 `~/.gitconfig`

What's less clear to me right now (I had a whole big thing typed up ...) is how profile and branch settings should interact; and which settings use which fallbacks and when, and so on...

**Comment by jnsnow** [comment]
No, sorry, I was mistaken about the current behavior... I was having a similar problem for an unrelated reason!

The way the include directives work is that it inlines them immediately; I think this means that we want to list them in order from least to most relevant, so:

```
[include]                                                                                                                                         
    path = ~/.gitconfig                                                                                                                           
    path = .gitpublish
    path = .git/config
```

I think that the current order actually does mean that ~/.gitconfig is going to bowl over everything else; which is probably not what we wanted....?

**Comment by ehabkost** [comment]
I was considering sending a patch to change the order and place `~/.gitconfig` at lowest priority, but I think this seems trickier than I've thought. I was looking only at `get_profile_var()`, but we do have options that may come from `git-config` directly, and the rules for each variable are slightly different from the others. I have enumerated all `get_profile_var()` and `get_profile_var_list()` callers below, for reference.

Then, we have two other `git_config_with_profile()` callers:

* `check_profile_exists():` this one is indifferent, as it just checks if the option exists. Good.
* `get_first_profile():` this one is tricky: the **first** profile is the one that appears in the **lowest priority** config file (!).  Probably not the behavior we want.

My question is: if we have multiple profiles in `~/.gitconfig`, `.gitpublish` and `.git/config`, which one should be the default?

--------

Raw list of `get_profile_var()` and `get_profile_var_list()` callers below:

* `base`:
```python
    base = options.base
    if not base:
        base = git_get_config('branch', current_branch, 'gitpublishbase')
    if not base:
        base = get_profile_var(options.profile_name, 'base')
    if not base:
        base = git_get_config('git-publish', 'base')
    if not base:
        base = 'master'
```
* `to`:
```python
    to = set([to_text(_) for _ in options.to])
    if not options.edit and not options.override_to:
        to = to.union(git_get_config_list('branch', topic, 'gitpublishto'))
        to = to.union(get_profile_var_list(options.profile_name, 'to'))
```
* `cc`:
```python
    cc = set([to_text(_) for _ in options.cc])
    if not options.edit and not options.override_cc:
        cc = cc.union(git_get_config_list('branch', topic, 'gitpublishcc'))
        cc = cc.union(get_profile_var_list(options.profile_name, 'cc'))
```
* `cc_cmd`:
```python
    cc_cmd = options.cc_cmd
    if not cc_cmd:
        cc_cmd = git_get_config('branch', topic, 'gitpublishcccmd') or \
                 get_profile_var(options.profile_name, 'cccmd')
```
* `blurb_template`:
```python
    blurb_template = options.blurb_template
    if not blurb_template:
        blurb_template = '\n'.join(get_profile_var_list(options.profile_name, 'blurb-template'))
    if not blurb_template:
        blurb_template = "*** BLURB HERE ***"
```
* `remote`:
```python
    if options.pull_request:
        remote = git_get_config('branch', topic, 'pushRemote')
        if remote is None:
            remote = git_get_config('remote', 'pushDefault')
        if remote is None:
            remote = git_get_config('branch', topic, 'remote')
        if remote is None or remote == '.':
            remote = get_profile_var(options.profile_name, 'remote')
        if remote is None:
            print('''Unable to determine remote repo to push.  Please set git config
branch.%s.pushRemote, branch.%s.remote, remote.pushDefault, or
gitpublishprofile.%s.remote''' % (topic, topic, options.profile_name))
            return 1
```
* `check_url`:
```python
        check_url = options.check_url
        if check_url is None:
            check_url_var = get_profile_var(options.profile_name, 'checkUrl')
            if check_url_var is None:
                check_url_var = git_get_config('git-publish', 'checkUrl')
            if check_url_var is not None:
                check_url = bool_from_str(check_url_var)
        if check_url is None:
            check_url = True
```
* `sign_pull`:
```python
        sign_pull = options.sign_pull
        if sign_pull is None:
            sign_pull_var = get_profile_var(options.profile_name, 'signPull')
            if sign_pull_var is None:
                sign_pull_var = git_get_config('git-publish', 'signPull')
            if sign_pull_var is not None:
                sign_pull = bool_from_str(sign_pull_var)
        if sign_pull is None:
            sign_pull = True
```
* `message`:
```python
    profile_message_var = get_profile_var(options.profile_name, 'message')
    if options.message is not None:
        message = options.message
    elif git_get_tag_message(tag_name_staging(topic)):
        # If there is a staged tag message, we definitely want a cover letter
        message = True
    elif profile_message_var is not None:
        message = bool_from_str(profile_message_var)
    elif options.pull_request:
        # Pull requests always get a cover letter by default
        message = True
    else:
        config_cover_letter = git_get_config('format', 'coverLetter')
        if config_cover_letter is None or config_cover_letter.lower() == 'auto':
            # If there are several commits we probably want a cover letter
            message = get_number_of_commits(base) > 1
        else:
            message = bool_from_str(config_cover_letter)
```
* `keyid`:
```python
    keyid = options.keyid
    if keyid is None:
        keyid_var = get_profile_var(options.profile_name, 'signingkey')
        if keyid_var is None:
            keyid_var = git_get_config('git-publish', 'signingkey')
```
* `prefix`:
```python
    prefix = options.prefix
    if prefix is not None:
        git_set_config('branch', topic, 'gitpublishprefix', prefix)
    else:
        prefix = git_get_config('branch', topic, 'gitpublishprefix')
    if prefix is None:
        prefix = get_profile_var(options.profile_name, 'prefix')
    if prefix is None:
        if options.pull_request:
            prefix = 'PULL'
        else:
            prefix = git_get_config('format', 'subjectprefix') or 'PATCH'
    if number > 1:
        prefix = '%s v%d' % (prefix, number)
```
* `suppress_cc`:
```python
        suppress_cc = options.suppress_cc
        if suppress_cc is None:
            suppress_cc = get_profile_var(options.profile_name, 'suppresscc')
```
* `signoff`:
```python
        if options.signoff:
            signoff = True
        else:
            signoff = get_profile_var(options.profile_name, 'signoff')
```
* `inspect_emails`:
```python
        if options.inspect_emails:
            inspect_emails = True
        else:
            inspect_emails = get_profile_var(options.profile_name, 'inspect-emails')
```
* `notes`:
```python
        if options.notes:
            notes = True
        else:
            notes = get_profile_var(options.profile_name, 'notes')
```

**Comment by ehabkost** [comment]
> My question is: if we have multiple profiles in `~/.gitconfig`, `.gitpublish` and `.git/config`, which one should be the default?

The answer is already in the code: `default` is always the default profile. The only caller of `get_first_profile()` doesn't care at all which profile it is:

```python
    if not check_profile_exists(options.profile_name):
        if options.profile_name == 'default':
            if get_first_profile():
                print('Using defaults when a non-default profile exists. Forgot to pass --profile ?')
        else:
            print('Profile "%s" does not exist, please check .gitpublish or git-config(1) files' % options.profile_name)
            return 1
```

I will send a patch to rename it to `has_profiles()`, making the ordering of profiles not matter at all.

**Comment by ehabkost** [comment]
Fixed by #67.
